### PR TITLE
FFZ should explicit use 4x emote links, also fix scale factors

### PR DESCRIPTION
**Pull request by goldbattle**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

This fixes FFZ to have the correct scale for their 4x higher resolution emotes.
This is also in preparation for 7tv support which has higher resolution emotes.

Comments:
- Scale factor should be ration to the base size of 32px
- Example: 32 / 112 = 0.286 scale factor for 3x emotes
- Example: 32 / 128 = 0.250 scale factor for 4x emotes

### CODE REVIEW

**Comment by goldbattle** [comment]
Any updates on what is blocking this? If nothing, I can fix the conflicts so we can merge.

File: src/providers/bttv/BttvEmotes.cpp [hunk 1]
```diff
@@ -94,7 +94,7 @@ namespace {
                     ImageSet{
                         Image::fromUrl(getEmoteLinkV3(id, "1x"), 1),
                         Image::fromUrl(getEmoteLinkV3(id, "2x"), 0.5),
-                        Image::fromUrl(getEmoteLinkV3(id, "3x"), 0.25),
+                        Image::fromUrl(getEmoteLinkV3(id, "3x"), 0.286),
```
Comment (pajlada) [inline src/providers/bttv/BttvEmotes.cpp:97]: The BTTV 3x emote is actually 4x-scaled of the 1x-version

Take KKona for example
```sh
for size in {1,2,3}; do echo "$size"; curl -s https://cdn.betterttv.net/emote/566ca04265dbbdab32ec054a/${size}x | file -; done
1
/dev/stdin: PNG image data, 25 x 34, 8-bit/color RGBA, non-interlaced
2
/dev/stdin: PNG image data, 50 x 68, 8-bit/color RGBA, non-interlaced
3
/dev/stdin: PNG image data, 100 x 136, 8-bit/color RGBA, non-interlaced
```

Then KKoooona from FFZ
```sh
for size in {1,2,4}; do echo "$size"; curl -s https://cdn.frankerfacez.com/emoticon/164405/$size | file -; done
1
/dev/stdin: PNG image data, 80 x 20, 8-bit/color RGBA, non-interlaced
2
/dev/stdin: PNG image data, 160 x 40, 8-bit/color RGBA, non-interlaced
4
/dev/stdin: PNG image data, 320 x 80, 8-bit/color RGBA, non-interlaced
```

The ratios are the same yet they're called differently


7TV's supports all 4 versions and the scales are all correct
```sh
for size in {1,2,3,4}; do echo "$size"; curl -s https://cdn.7tv.app/emote/603cac391cd55c0014d989be/${size}x.png | file -; done
1
/dev/stdin: PNG image data, 42 x 32, 8-bit/color RGBA, non-interlaced
2
/dev/stdin: PNG image data, 84 x 64, 8-bit/color RGBA, non-interlaced
3
/dev/stdin: PNG image data, 126 x 96, 8-bit/color RGBA, non-interlaced
4
/dev/stdin: PNG image data, 168 x 128, 8-bit/color RGBA, non-interlaced
```

I believe the correct change here would be to omit the 3x version, and treat BTTV's 3x version as a 4x version

**Review by pajlada — CHANGES_REQUESTED** [review]
Thanks for the bump - I had lost track of this simple PR

File: src/providers/bttv/BttvEmotes.cpp (hunk 1 above)
Comment (Nerixyz) [inline src/providers/bttv/BttvEmotes.cpp:97]: I don't think `0.286` is the right value (at least not for all emotes). For `catJAM` it's `28/112 = 0.25` and `KKona` it's `25/100 = 0.25`.

**Comment by Nerixyz** [comment]
I made some side-by-side tests (is this the expected result?). 
Emotes: [PagMan (BTTV)](https://betterttv.com/emotes/5f0ed97873e8e20538d812a6), Kappa, [FeelsOkayMan (FFZ)](https://www.frankerfacez.com/emoticon/145947-FeelsOkayMan), PogChamp

| Title | Result |
|---|---|
| this PR, zoom: 2.1x |  |
| before, zoom: 2.1x |  |

**Comment by goldbattle** [comment]
Looks like I need to spend a bit more time to investigate, converted this PR to a draft for now.

**Comment by Nerixyz** [comment]
> Looks like I need to spend a bit more time to investigate, converted this PR to a draft for now.

You don't need to do everything in one PR. You could do the 4x images in this PR and investigate scale factors in the next. This would also make it easier to review.

**Comment by goldbattle** [comment]
So I collected some statistics about the emote size for different resolutions here is what I found.
width_vs_height_hist

The emotes for the histogram are "binned" to group emotes that are near in size to get a general idea. The number in the image are the number of emotes that are in that bin.
```
DATA: 54509 bttv emotes loaded
    1 -> xbin = 3.50px | ybin = 1.30px | 54508 emotes
    2 -> xbin = 6.95px | ybin = 2.60px | 54509 emotes
    3 -> xbin = 13.95px | ybin = 5.25px | 54508 emotes
DATA: 19723 ffz emotes loaded
    1 -> xbin = 5.15px | ybin = 2.40px | 19723 emotes
    2 -> xbin = 10.30px | ybin = 4.85px | 19723 emotes
    4 -> xbin = 20.60px | ybin = 9.70px | 19723 emotes
```


The main conclusion I can see are the following:
- BTTV in general has much lower variance in height compared to FFZ which has two grouping of emotes
- FFZ has a lot of common emotes that are around 30px wide, and another bunch that are 35px wide
- BTTV wide emotes are in general around ~28px tall
- FFZ wide emotes are a little shorter at ~22px tall
- The 3x BTTV are ~110px high which is what we expect for 3x emotes 128px?
- FFZ's 4x are ~110px and ~128px (see the two groups in the bottom right), which seems to indicate that they have a mix of short and correct height 4x emotes?





This is the scatter plot version:
width_vs_height_scatter

**Comment by pajlada** [comment]
If we only want to support 3x image variants, then I believe prior to this PR that FFZ, BTTV, and Twitch emotes are all correctly scaled looking at the max 'average' emote

For FFZ that would be 32px-height emotes
[1](https://cdn.frankerfacez.com/emoticon/230495/1): 32 (loaded as image1x, scale 1.0)
[2](https://cdn.frankerfacez.com/emoticon/230495/2): 64 (loaded as image2x, scale 0.5)
[4](https://cdn.frankerfacez.com/emoticon/230495/4): 128 (loaded as image3x, scale 0.25)

For BTTV that would be 28px-height emotes
[1x](https://cdn.betterttv.net/emote/5e8b9349518d2b5ce0b0f335/1x): 28 (loaded as image1x, scale 1.0)
[2x](https://cdn.betterttv.net/emote/5e8b9349518d2b5ce0b0f335/2x): 56 (loaded as image2x, scale 0.5)
[3x](https://cdn.betterttv.net/emote/5e8b9349518d2b5ce0b0f335/3x): 112 (loaded as image3x, scale 0.25)

For Twitch that would be 28px-height emotes
[1.0](https://static-cdn.jtvnw.net/emoticons/v2/80481/default/dark/1.0): 28 (loaded as image1x, scale 1.0)
[2.0](https://static-cdn.jtvnw.net/emoticons/v2/80481/default/dark/2.0): 56 (loaded as image2x, scale 0.5)
[3.0](https://static-cdn.jtvnw.net/emoticons/v2/80481/default/dark/3.0): 112 (loaded as image3x, scale 0.25)

all boiling down to FFZ emotes being at most 32px tall, BTTV emotes being at most 28px tall, Twitch emotes being at most 28px tall (discounting outliers like NaM or freak old uploads)

I think looking at scale factors & adding a fourth emote link variants must be done separately for this to get the care it deserves, as adding a fourth emote link means we'd need to decide where to put it. ould it go as a 4x link? or as a 2.5x link inbetween 2x and 3x?

**Comment by pajlada** [comment]
Closing as I don't think this is something we necessarily want to change - happy to have a discussion about it in here or elsewhere!

**Pull request closed without merging**